crypto/blake2/blake2b.c

```diff
@@ -93,7 +92,54 @@ int BLAKE2b_Init(BLAKE2B_CTX *c)
     memset(P->reserved, 0, sizeof(P->reserved));
     memset(P->salt,     0, sizeof(P->salt));
     memset(P->personal, 0, sizeof(P->personal));
+}
+
+void blake2b_param_set_digest_length(BLAKE2B_PARAM *P, uint8_t outlen)
+{
+    P->digest_length = outlen;
+}
+
+void blake2b_param_set_key_length(BLAKE2B_PARAM *P, uint8_t keylen)
+{
+    P->key_length = keylen;
+}
+
+void blake2b_param_set_personal(BLAKE2B_PARAM *P, const uint8_t personal[BLAKE2B_PERSONALBYTES])
+{
+    memcpy(P->personal, personal, sizeof(P->personal));
+}
+
+void blake2b_param_set_salt(BLAKE2B_PARAM *P, const uint8_t salt[BLAKE2B_SALTBYTES])
+{
+    memcpy(P->salt, salt, sizeof(P->salt));
+}
+
+/*
+ * Initialize the hashing context with the given parameter block.
+ * Always returns 1.
+ */
+int BLAKE2b_Init(BLAKE2B_CTX *c, const BLAKE2B_PARAM *P)
+{
+    blake2b_init_param(c, P);
+    return 1;
+}
+
+/*
+ * Initialize the hashing context with the given parameter block and key.
+ * Always returns 1.
+ */
+int BLAKE2b_Init_key(BLAKE2B_CTX *c, const BLAKE2B_PARAM *P, const void *key)
+{
     blake2b_init_param(c, P);
+
+    /* Pad the key to form first data block */
+    {
+        uint8_t block[BLAKE2B_BLOCKBYTES] = {0};
+        memcpy(block, key, P->key_length);
```

Comment:
Should there be a check somewhere that ensures `P->key_length` doesn't exceed `BLAKE2B_BLOCKBYTES`?

Reply:
All `BLAKE2_Init` and `blake2_param` are lower level functions that do not check the validity of their parameters, the validation logic is in `blake2b_mac_ctrl`, which is at this point the only way for the user to set parameter values.
I will add some clarification in the comments.